test/functional/p2p_node_network_limited.py

```diff
@@ -46,6 +50,65 @@ def setup_network(self):
         self.add_nodes(self.num_nodes, self.extra_args)
         self.start_nodes()
 
+    def test_avoid_requesting_historical_blocks(self):
+        self.log.info("Test full node does not request blocks beyond the limited peer threshold")
+        pruned_node = self.nodes[0]
+        miner = self.nodes[1]
+        full_node = self.nodes[2]
+
+        # Connect and generate block to ensure IBD=false
+        self.connect_nodes(1, 0)
+        self.connect_nodes(1, 2)
+        self.generate(miner, 1)
+
+        # Verify peers are out of IBD
+        for node in self.nodes:
+            assert not node.getblockchaininfo()['initialblockdownload']
+
+        # Isolate full_node (the node will remain out of IBD)
+        full_node.setnetworkactive(False)
+        self.wait_until(lambda: len(full_node.getpeerinfo()) == 0)
+
+        # Mine blocks and sync the pruned node. Surpass the NETWORK_NODE_LIMITED threshold.
+        self.generate(miner, NODE_NETWORK_LIMITED_MIN_BLOCKS, sync_fun=self.no_op)
+        self.sync_blocks([miner, pruned_node])
+
+        # Connect full_node to prune_node and check peers don't disconnect right away.
+        # (they will disconnect if full_node, which is chain-wise behind, request blocks
+        # older than NODE_NETWORK_LIMITED_MIN_BLOCKS)
+        start_height_full_node = full_node.getblockcount()
+        full_node.setnetworkactive(True)
+        self.connect_nodes(2, 0)
+        assert_equal(len(full_node.getpeerinfo()), 1)
+
+        # Wait until the full_node is headers-wise sync
+        best_block_hash = pruned_node.getbestblockhash()
+        self.wait_until(lambda: next(filter(lambda x: x['hash'] == best_block_hash, full_node.getchaintips()))['status'] == "headers-only")
+
+        # Now, since the node aims to download a window of 1024 blocks,
+        # ensure it requests the blocks below the threshold only (with a
+        # 2-block buffer). And also, ensure it does not request blocks
+        # deeper than the threshold (minus the buffer).
+        limited_threshold_buffer = 2
+
+        # Traverse in reverse order, ensuring the node stores all blocks
+        # expected to be downloaded from the limited peer.
+        tip_height = pruned_node.getblockcount()
+        self.wait_until(lambda: not try_rpc(-1, "Block not found", full_node.getblock, pruned_node.getbestblockhash()))
+        for height in range(tip_height, start_height_full_node + limited_threshold_buffer, -1):
+            full_node.getblock(pruned_node.getblockhash(height))  # just need to assert it does not raise an exception
+
+        # Verify all blocks that should not have been requested.
+        for move in range(1, limited_threshold_buffer):
+            height = start_height_full_node + move
+            assert_raises_rpc_error(-1, "Block not found on disk", full_node.getblock, pruned_node.getblockhash(height))
```

Comment:
The full node is at height `1` and the pruned node is at height `289`. The first loop verifies that the full node has blocks `[4, 289]` (`286` blocks in total), which is ok. But then it only checks that the full node does not have block `2`. What about block `3`? It is not checked and should not have been downloaded but it has been downloaded.

To avoid the gap the range in the second loop should be `range(1, limited_threshold_buffer + 1)` or even:

```python
for height in range(start_height_full_node + 1, start_height_full_node + 1 + limited_threshold_buffer):
```

or have one loop instead of two:

```python
for height in range(start_height_full_node + 1, tip_height + 1):
    if height <= tip_height - (NODE_NETWORK_LIMITED_MIN_BLOCKS - 2):
        assert_raises_rpc_error(-1, "Block not found on disk", full_node.getblock, pruned_node.getblockhash(height))
    else:
        full_node.getblock(pruned_node.getblockhash(height))
```

Reply:
Done as suggested. Thanks.